### ISSUE TITLE
no-unused-expressions rule should allow `foo && foo()`

### ISSUE THREAD

**Issue by KevinGrandon** [eslint/eslint#2684]
From the rule docs, `This rule aims to eliminate unused expressions. The value of an expression should always be used, except in the case of expressions that side effect: function calls, assignments, and the new operator.`.

I believe the syntax ``foo && foo()` should be considered valid here as it's a function call with potential side effects. This syntax allows for quite clean truthfulness checks inside code, and it would be nice if I could use this syntax with the rest of the rules inside no-unused-expressions. (If not, please let me opt out of just this one use-case)

**Comment by KevinGrandon** [comment]
Added here: https://github.com/eslint/eslint/commit/6a4059392cb2d1018e57c6d7debfc319ecd4f3c2

**Comment by michaelficarra** [comment]
Duplicate of https://github.com/eslint/eslint/issues/2418.

**Comment by ilyavolodin** [comment]
@KevinGrandon as stated in #2418 and in the text that you quoted above this rule is designed to flag expression where the value is unused. In the example you provided, the output of the function as well as the result of check is lost. This is exactly what this rule is aiming at flagging. If the rule doesn't fit with your coding style, you should just turn that rule off, or disable it inline if you only have a few exceptions to the rule.

**Comment by KevinGrandon** [comment]
@ilyavolodin - How would I submit a feature request then to split this check out from `no-unused-expressions`?

I think you should be able to have something which warns on `"Hello world";`, but does not warn on `foo && foo()`.

**Comment by ilyavolodin** [comment]
@KevinGrandon Unfortunately this has been discussed many times before, and in each case the decision was not to change the way this rule works. So I do not think we will make this change, unless some new and very persuasive arguments are made.
Your best bet is to either disable this rule inline for the cases that you want to allow, or disable this rule all together and create a plugin with a clone of this rule, that would do what you want.

**Comment by jwarkentin** [comment]
The obvious solution is to create a separate rule that differentiates between all unused expressions versus noops. Any time there is a function call or an assignment as part of an expression, the rule would not flag it as a noop. All other unused expressions would be caught by the rule.

This is obviously a rule that the community would like and it has been demonstrated many times that the pattern is both common and not harmful. There is clearly a case to be made for differentiating between types of unused expressions. Actually, I think it's incorrect to call it an "unused expression" as it is clearly being used, just not assigned. It is still very useful in JavaScript.

I would be more than happy to turn off this rule if there was a replacement that would still catch noops like:

```js
a;
a + b;
"Hello World!";
```

As stated before, function calls and assignments are not noops and are very useful as part of an unassigned expression.

**Comment by KevinGrandon** [comment]
+1. I recommend creating a new rule, `no-usless-expresions` as well, which would have a subset of the original functionality of the `no-unused-expressions` rule. I think this is the best course of action for the core project, so I will file an issue for a feature request and see what happens.
